tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs

```diff
@@ -481,6 +481,90 @@ public void DescriptionChanged_should_be_raised_when_the_description_changes()
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27018)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27019)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27020)));
+            _capturedEvents.Clear();
+
+            for (int i = 0; i < 3; i++)
+            {
+                var selectedServer = SelectServerAttempt(
+                    subject,
+                    new DelegateServerSelector((c, s) => s), // do not filter servers
+                    async);
+
+                var selectedServerPort = ((DnsEndPoint)selectedServer.EndPoint).Port;
+                selectedServerPort.Should().Be(27020);
+                _capturedEvents.Next().Should().BeOfType<ClusterSelectingServerEvent>();
+                _capturedEvents.Next().Should().BeOfType<ClusterSelectedServerEvent>();
+            }
+
+            numberOfCustomServerSelectorCalls.Should().Be(3);
+            _capturedEvents.Any().Should().BeFalse();
+        }
+
+        [Theory]
+        [ParameterAttributeData]
+        public void SelectServer_should_call_custom_selector_if_there_are_no_eligible_servers([Values(true, false)] bool async)
+        {
+            int numberOfCustomServerSelectorCalls = 0;
+            var customServerSelector = new DelegateServerSelector((c, s) =>
+            {
+                numberOfCustomServerSelectorCalls++;
+                return s; // return all servers
+            });
+
+            var settings = _settings.With(
+                postServerSelector: customServerSelector,
+                serverSelectionTimeout: TimeSpan.FromSeconds(1));
+            var subject = new StubCluster(settings, _mockServerFactory.Object, _capturedEvents);
+
+            subject.Initialize();
+            subject.SetServerDescriptions(
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27017)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27018)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27019)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27020)));
+            _capturedEvents.Clear();
+
+            for (int i = 0; i < 3; i++)
+            {
+                var exception = Record.Exception(
+                    () =>
+                        SelectServerAttempt(
+                            subject,
+                            new DelegateServerSelector((c, s) => new ServerDescription[0]), // no eligible servers
+                            async));
+
+                exception.Should().BeOfType<TimeoutException>();
```

Comment:
Isn't this going to be a very slow test if it has to wait 30 seconds for a timeout?

And it's going to be even slower if we run it 3 times.